### PR TITLE
CS-7 Admin CRUD

### PR DESCRIPTION
**Pull request by Akshay-Chhikara**

### CODE REVIEW

File: app/controllers/admins/sessions_controller.rb [hunk 1]
```diff
@@ -5,10 +5,10 @@ class Admins::SessionsController < Devise::SessionsController
 
   def check_subdomain?
     subdomain = request.subdomain
-    if subdomain == 'www'
-      redirect_to root_url
+    if ['www', 'ftp', 'ssh'].include?(subdomain)
```
Comment (yuktikhurana) [inline app/controllers/admins/sessions_controller.rb]: Let's a validation for exclusion in the company model

File: app/controllers/admins/sessions_controller.rb [hunk 2]
```diff
@@ -5,10 +5,10 @@ class Admins::SessionsController < Devise::SessionsController
...
+      return redirect_to root_url(host: Rails.application.config.action_mailer.default_url_options[:host])
     elsif subdomain.blank? || Company.find_by(subdomain: subdomain).nil?
-      redirect_to new_admin_registration_url(host: Rails.application.config.action_mailer.default_url_options[:host]), alert: 'Get Registered First'
+      return redirect_to new_company_url(host: Rails.application.config.action_mailer.default_url_options[:host]), alert: 'Get Registered First'
```
Comment (yuktikhurana) [inline app/controllers/admins/sessions_controller.rb]: No need to write return here as we're redirecting

File: app/controllers/admins_controller.rb [hunk 3]
```diff
@@ -3,7 +3,50 @@ class AdminsController < ApplicationController
...
 
   def index
+  end
+
+  def new
+    @admin = Admin.new
+  end
+
+  def create
+    @company = Company.where(id: current_admin.company.id).first
```
Comment (yuktikhurana) [inline app/controllers/admins_controller.rb]: @company is not required

File: app/controllers/admins_controller.rb [hunk 4]
```diff
@@ -3,7 +3,50 @@ class AdminsController < ApplicationController
...
+    @admin = @company.admins.build(admin_params)
+    if @admin.save_without_confirmation
+      redirect_to admins_path, notice: 'New agent added'
+    else
+      render :new
+    end
+  end
+
+  def edit
+    @admin = current_admin
```
Comment (yuktikhurana) [inline app/controllers/admins_controller.rb]: @admin not needed. current_admin will serve the purpose. it is also available in views

File: app/controllers/admins_controller.rb [hunk 5]
```diff
@@ -3,7 +3,50 @@ class AdminsController < ApplicationController
...
+  end
+
+  def update
     @admin = current_admin
+    @admin.name = name_param[:name]
```
Comment (yuktikhurana) [inline app/controllers/admins_controller.rb]: we can use  admin_params[:name]. Also we have a method update for updating

File: app/controllers/admins_controller.rb [hunk 6]
```diff
@@ -3,7 +3,50 @@ class AdminsController < ApplicationController
...
+      render :new
+    end
+  end
+
+  def edit
+    @admin = current_admin
+  end
+
+  def update
     @admin = current_admin
```
Comment (yuktikhurana) [inline app/controllers/admins_controller.rb]: use DRY

File: app/controllers/companies_controller.rb [hunk 7]
```diff
@@ -1,2 +1,33 @@
 class CompaniesController < ApplicationController
+
+  def new
+    @company = Company.new
+    @admin = @company.admins.build
+  end
+
+  def create
+    @company = Company.new(company_params)
+    @admin = @company.admins.build(admin_params)
```
Comment (yuktikhurana) [inline app/controllers/companies_controller.rb]: use dry

File: app/mailers/admin_mailer.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+class AdminMailer < ActionMailer::Base
+  helper :subdomain
+  default template_path: 'devise/mailer'
```
Comment (yuktikhurana) [inline app/mailers/admin_mailer.rb]: Use admin mailer path

File: app/mailers/admin_mailer.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  default from: 'CServe'
+
+  def set_password_instructions(admin, token)
+    @admin = admin
+    @token = token
+    mail(to: @admin.email, subject: 'Set Password for your account')
```
Comment (yuktikhurana) [inline app/mailers/admin_mailer.rb:5]: subject: please confirm with the PM

File: app/models/admin.rb [hunk 10]
```diff
@@ -9,4 +7,17 @@ class Admin < ActiveRecord::Base
   def subdomain
     company.subdomain
```
Comment (yuktikhurana) [inline app/models/admin.rb]: use delegate

File: app/views/admins/_error_alert.html.erb [hunk 11]
```diff
@@ -0,0 +1,6 @@
+<% if @admin.errors.any? %>
```
Comment (yuktikhurana) [inline app/views/admins/_error_alert.html.erb]: reuse the existing error partial

File: app/views/admins/_form.html.erb [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    </div>
+  </div>
+  <div class='form-group'>
+    <%= f.label :name, 'Admin name*', class:'col-sm-4 control-label' %>
+    <div class='col-sm-8'>
+      <%= f.text_field :name, class: 'form-control', :placeholder => "Admin Name" %>
+    </div>
+  </div>
+  <div style="display:none">
+    <%= f.hidden_field :role, value: 'super_admin' %>
```
Comment (yuktikhurana) [inline app/views/admins/_form.html.erb]: already a hidden field. Why display none ? Also why do we need this role to be passed from the form ? We can set it in the model itself as currently we're not managing any roles or maybe just leave it empty. Remove validation from role

File: app/controllers/admins/sessions_controller.rb (hunk 2 above)
Comment (aditya-kapoor) [inline app/controllers/admins/sessions_controller.rb]: :+1:

**Pull request closed without merging**